src/gennylib/src/context.cpp

```diff
@@ -133,8 +140,13 @@ DefaultRandom& WorkloadContext::getRNGForThread(ActorId id) {
     if (this->isDone()) {
         BOOST_THROW_EXCEPTION(std::logic_error("Cannot create RNGs after setup"));
     }
+
+    bool success = false;
+
+    std::lock_guard<std::mutex> lk(_rngLock);
     if (auto rng = _rngRegistry.find(id); rng == _rngRegistry.end()) {
-        auto [it, success] = _rngRegistry.try_emplace(id, _rng());
+        auto [it, returnedSuccess] = _rngRegistry.try_emplace(id, _rngSeed + std::hash<long>{}(id));
```

Comment:
Copying forward the discussion
This doesn't jump as far as the previous version for each RNG, but my understanding is that for a decent modern RNG this should be fine?

 
Member
@rtimmons rtimmons 7 days ago
This is clever. My instinct says to do a std::hash or something of the id rather than just straight addition, but I'd actually bring in David Daly here - he did some pretty extensive investigations into proper RNGs with his initial WLG proto-genny concept a few years ago.

Either way, this will cause workload behavior to change since Actors will get different seeds after this change. I'm not sure of the implications there but it may require a bit of optics-management when this goes out.

 
Member
Author
@aggrand aggrand 14 minutes ago
Ended up using a hash (not visible here since it's treating this as outdated). @dalyd any thoughts on whether that's random enough, or if there's a different approach we should use?

Let me see if I understand the issue. I'm guessing because we are creating these in parallel we can't call _rng() because it is singe threaded, so we are trying to find a way to get nicely distributed seeds for the random number generators. Is that correct? 

Assuming it is a few thoughts. 
1. Could we just generate the random seeds and put them in an array before parallelizing? That would allow (possibly) doing rng creation in parallel, but still using the existing seeding code, and would lead to the same per thread behavior before and after this change. Otherwise we do get a change in workload behavior, although hopefully an insignificant one. 
2. I'm not sure how sound my original code was, that seeded each thread with a random number from another rng. I'm going to dig and poke a bit.

Reply:
After further review I feel better about the original code, although I wish I had passed more random state into the constructor (longer seed sequence)

Reply:
Yeah pretty much, and if we make `_rng()` naively thread-safe it would be nondeterministic which thread gets which seed.

I didn't want to do option 1 because we don't necessarily want to assume how many actors we have to start with (would rather leave that up to the `ActorProducer`s).

One option would be, if actor if id `n` requests an rng, we produce all the rngs up to `n` using the existing code, then anyone else requesting an rng takes that. I guess everyone would be waiting if the last actor manages to request first, but they also don't have to construct anything themselves so maybe that balances out? (Not that initial rng generation was large in the flame graphs anyway.)

@dalyd if a workload behavior change would be bad I can take that approach instead! I'd assumed it would be okay if we change it one-off but if we want to avoid it at all I can take the above-mentioned approach

Reply:
Yes, could we give that a try? Let's see if it adds up or not. And note only the seed generation part needs to be serialized.